crates/miden-lib/src/note/well_known_note.rs

```diff
@@ -145,4 +147,101 @@ impl WellKnownNote {
+                };
+
+                // check that the account ID in the note inputs equal to the target account ID
+                if input_account_id == target_account_id {
+                    NoteAccountCompatibility::Yes
+                } else {
+                    NoteAccountCompatibility::No
+                }
+            },
+            WellKnownNote::P2IDR => {
+                let note_inputs = note.inputs().values();
+                if note_inputs.len() != 3 {
+                    return NoteAccountCompatibility::No;
+                }
+
+                let recall_height: Result<u32, _> = note_inputs[2].try_into();
+                // Return `No` if the note input value which represents the recall height is invalid
+                let Ok(recall_height) = recall_height else {
+                    return NoteAccountCompatibility::No;
+                };
+
+                // Return `No` if the note input values used to construct the account ID are invalid
+                let Some(input_account_id) = try_read_account_id_from_inputs(note_inputs) else {
+                    return NoteAccountCompatibility::No;
+                };
+
+                if block_ref.as_u32() >= recall_height {
+                    let sender_account_id = note.metadata().sender();
+                    // if the sender can already reclaim the assets back, then:
+                    // - target account ID could be equal to the inputs account ID if the note is
+                    //   going to be consumed by the target account
+                    // - target account ID could be equal to the sender account ID if the note is
+                    //   going to be consumed by the sender account
+                    if [input_account_id, sender_account_id].contains(&target_account_id) {
+                        NoteAccountCompatibility::Yes
+                    } else {
+                        NoteAccountCompatibility::No
+                    }
+                } else {
+                    // in this case note could be consumed only by the target account
+                    if input_account_id == target_account_id {
+                        NoteAccountCompatibility::Yes
+                    } else {
+                        NoteAccountCompatibility::No
+                    }
+                }
+            },
+            WellKnownNote::SWAP => {
+                if note.inputs().values().len() != 10 {
+                    return NoteAccountCompatibility::No;
+                }
+
+                NoteAccountCompatibility::Maybe
+            },
+        }
+    }
+}
+
+// HELPER FUNCTIONS
+// ================================================================================================
+
+/// Reads the account ID from the note inputs.
+///
+/// Returns None if the note input values used to construct the account ID are invalid.
+fn try_read_account_id_from_inputs(note_inputs: &[Felt]) -> Option<AccountId> {
```

Comment:
nit: I would specify in the comments that we try to read the ID from the first 2 inputs.